legacy/builder/container_find_includes.go

```diff
@@ -144,16 +153,69 @@ func (s *ContainerFindIncludes) findIncludes(ctx *types.Context) error {
 		queueSourceFilesFromFolder(ctx, sourceFilePaths, sketch, srcSubfolderPath, true /* recurse */)
 	}
 
-	for !sourceFilePaths.Empty() {
-		err := findIncludesUntilDone(ctx, cache, sourceFilePaths.Pop())
+	preloadCachedFolder(ctx, cache)
+
+	// The first source file is the main .ino.cpp
+	// handle it first to setup environment for other files
+	findIncludesUntilDone(ctx, cache, sourceFilePaths.Pop())
+
+	var errorsList []error
+	var errorsMux sync.Mutex
+
+	queue := make(chan types.SourceFile)
+	job := func(source types.SourceFile) {
+		// Find all includes
+		err := findIncludesUntilDone(ctx, cache, source)
 		if err != nil {
-			cachePath.Remove()
-			return errors.WithStack(err)
+			errorsMux.Lock()
+			errorsList = append(errorsList, err)
+			errorsMux.Unlock()
+		}
+	}
+
+	// Spawn jobs runners to make the scan faster
+	var wg sync.WaitGroup
+	jobs := ctx.Jobs
+	if jobs == 0 {
+		jobs = runtime.NumCPU()
+	}
+	var unhandled int32 = 0
+	for i := 0; i < jobs; i++ {
+		wg.Add(1)
+		go func() {
+			for source := range queue {
+				job(source)
+				atomic.AddInt32(&unhandled, -1)
+			}
+			wg.Done()
+		}()
+	}
+
+	// Loop until all files are handled
+	for (!sourceFilePaths.Empty() || unhandled != 0 ) {
```

Comment:
Isn't this `unhandled` variable redundant here? the WaitGroup should already take care of waiting for all the processes to finish.

Reply:
The _unhandled_ is not redundant here. During the scanning process, we may find new source files. For example, suppose we have only 4 files to scan in the beginning, but we have 8 cores. Then the sourceFilePaths will be empty after the first goroutines fetch their task and the loop will be ended. But during the scanning  against the 4 files may append more new found files to the sourceFilePaths but they won't be handled. The variable _unhandled_ is to make sure sourceFilePaths is empty only after all goroutines finish the jobs.